compare_hashes/score.h

```diff
@@ -0,0 +1,60 @@
+#ifndef SCORE_H
+#define SCORE_H
+
+#include <cassert>
+
+typedef int location_in_submission;

```

Comment:
Why are primative types being `typedef`'d ? To someone reading code with these it removes all type information from a declaration. When I see something like a vector of `user_id` I would assume user_id is a class or struct not a string.

I think its better to have the variable name contain information to what its supposed to represent e.g:

```c++
std::unordered_map<std::string, std::pair<int, Score> >::iterator itr
std::vector<std::string> cheating_user_ids { "bitdiddle" };
```

is easier to read than

```c++
std::unordered_map<user_id, std::pair<int, Score> >::iterator user_id_itr
std::vector<user_id> cheating_user_ids { "bitdiddle" };
```

Reply:
The issue here is that we've changed the types of several of these in the past and it's easy to forget one or more places.  Also, if you see a type like `std::unordered_map<std::string, std::unordered_map<user_id, std::unordered_map<version_number, std::unordered_set<hash>>>>`.  That's pretty long, and can be confusing when you aren't sure which int means what.  It's hard to avoid such huge types when working with the data structures necessary for Lichen so we decided to typedef most things to make it easier to change types in the future and to make it easier to understand the structure of each data structure.

Reply:
Why would you not just typedef the container instead `std::string` or if the type gets really complicated abstract it away in a class or struct.

regardless I would recommend moving away from using a typedef of `user_id`  and other typedefs and just using the primitive, and I don't see how you can change the type in one place and forget it in others without getting compiler errors

Reply:
It's easy to change the type in one place and forget about it in others in cases where you're switching between numeric types of various kinds.  The original source of this was when we switched to using `unsigned int`s for the `version_number` because previously we had bounced around between `int`s, `char`s and a couple other types.  As far as creating a typedef for each container, we considered it but there are a lot of similar, but slightly different containers and it still doesn't resolve the issue of having multiple numeric types in a single data structure.  Creating objects for each structure would potentially be better but I'm not sure if it would be ideal due to the number of different types required.

Reply:
Switching between ints, chars, unsigned ints can be detected compile time in C++

You don't have to create an object for every single std container your using right now, abstracting some of the complexity away from these containers into classes can help you avoid these complicated type declarations 

typedefing a primitive doesn't solve your ultimate problem of communicating what these complicated containers are doing or why they're needed and makes the code harder to read